tools/testtool/main.go

```diff
@@ -0,0 +1,203 @@
+		Directory("/")
+
+	if *containerized {
+		return runContainerized(ctx, client, artifacts, config.KernelPath, config.QEMUCmd, flag.Args())
+	}
+
+	return runNatively(ctx, artifacts, config.KernelPath, config.QEMUCmd, flag.Args())
+}
+
+func runNatively(
+	ctx context.Context,
+	artifacts *dagger.Directory,
+	kpath, qemuCmd string,
+	args []string,
+) error {
+	tmp, err := os.MkdirTemp(".", "ci-testing")
+	if err != nil {
+		return fmt.Errorf("unable to create tmp dir: %w", err)
+	}
+	defer os.RemoveAll(tmp)
+
+	if ok, err := artifacts.Export(ctx, tmp); !ok || err != nil {
+		return fmt.Errorf("failed artifact export: %w", err)
+	}
+
+	pwd, err := os.Getwd()
+	if err != nil {
+		return fmt.Errorf("could not retrieve working directory: %w", err)
+	}
+
+	tmp, err = filepath.Abs(tmp)
+	if err != nil {
+		return fmt.Errorf("could not retrieve absolute path: %w", err)
+	}
+
+	kpath = filepath.Join(tmp, kpath)
+
+	qemuCmd = fmt.Sprintf(qemuCmd, filepath.Join(tmp, "zqemu", "pc-bios"))
+
+	// Rather than adding the QEMU Cmd to PATH in cmd.Env,
+	// we are doing this because args[0] can be qemu, and if that's the case,
+	// exec.Command does not evaluate the PATH in cmd.Env, but instead the one in the current environment.
+	// The PATH will also be restored after the program exits.
+	p := os.Getenv("PATH")
+	if err := os.Setenv("PATH", fmt.Sprintf("%s:%s", p, filepath.Join(tmp, "zqemu", "bin"))); err != nil {
+		return fmt.Errorf("failed to update PATH: %w", err)
+	}
+	defer os.Setenv("PATH", p)
+
+	cmd := exec.Command(args[0], args[1:]...)
+	cmd.Env = append(
+		os.Environ(),
+		fmt.Sprintf("UROOT_KERNEL=%s", kpath),
+		fmt.Sprintf("UROOT_QEMU=%s", qemuCmd),
+		fmt.Sprintf("UROOT_SOURCE=%s", pwd),
+		"UROOT_QEMU_COVERPROFILE=coverage.txt",
+		"UROOT_QEMU_TIMEOUT_X=7",
+	)
+	cmd.Stdin = os.Stdin
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return fmt.Errorf("failed execution: %w", err)
+	}
+
+	return nil
+}
+
+func runContainerized(
+	ctx context.Context,
+	client *dagger.Client,
+	artifacts *dagger.Directory,
+	kpath, qemuCmd string,
+	args []string,
+) error {
+	src := client.Host().Directory(".")
+
+	if ok, err := client.
+		Container().
+		From("golang:1.20").
+		WithMountedDirectory("/src", src).
+		WithMountedDirectory("/artifacts", artifacts).
+		WithWorkdir("/src").
+		WithEnvVariable("PATH", "/artifacts/zqemu/bin:$PATH", dagger.ContainerWithEnvVariableOpts{
+			Expand: true,
+		}).
+		WithEnvVariable("UROOT_KERNEL", filepath.Join("/artifacts", kpath)).
+		WithEnvVariable("UROOT_QEMU", fmt.Sprintf(qemuCmd, filepath.Join("/artifacts", "zqemu", "pc-bios"))).
+		WithEnvVariable("UROOT_SOURCE", "/src").
+		WithEnvVariable("UROOT_QEMU_COVERPROFILE", "coverage.txt").
+		WithEnvVariable("UROOT_QEMU_TIMEOUT_X", "7").
+		WithExec(args).
+		File("coverage.txt").
+		Export(ctx, "coverage.txt"); !ok || err != nil {
+		log.Fatalf("container error: %v", err)
```

Comment:
also, this should return an error rather than log Fatal.

the function right now either log.Fatals or returns nil, making the error useless. preferred to return the error.

Reply:
Done